File: src/main/java/kitchenpos/eatinorders/tobe/domain/OrderTable.java
package kitchenpos.eatinorders.tobe.domain;

import jakarta.persistence.*;
import kitchenpos.common.vo.PositiveNumber;

@Entity
@Table(name = "order_table")
public class OrderTable {

    @EmbeddedId
    private OrderTableId id;

    @Embedded
    private OrderTableName name;

    @Embedded
    private PositiveNumber numberOfGuests;

    @Column(name = "occupied", nullable = false)
    private boolean occupied;

    protected OrderTable() {
    }

    public static OrderTable create(String name) {
        return new OrderTable(
                OrderTableId.generate(),
                new OrderTableName(name),
                PositiveNumber.ZERO,
                false
        );
    }

    public OrderTable(String name, int numberOfGuests, boolean occupied) {
        this(OrderTableId.generate(), new OrderTableName(name), new PositiveNumber(numberOfGuests), occupied);
    }

    public OrderTable(OrderTableId id, OrderTableName name, PositiveNumber numberOfGuests, boolean occupied) {
        this.id = id;
        this.name = name;
        this.numberOfGuests = numberOfGuests;
        this.occupied = occupied;
    }

    public void sit() {
        this.occupied = true;
    }

    public void clear() {
        this.occupied = false;
        this.numberOfGuests = PositiveNumber.ZERO;
    }

    public void changeNumberOfGuests(int number) {
        changeNumberOfGuests(new PositiveNumber(number));
    }

    public void changeNumberOfGuests(PositiveNumber number) {
        this.numberOfGuests = number;
    }

    public OrderTableId getId() {
        return id;
    }

    public OrderTableName getName() {
        return name;
    }

    public PositiveNumber getNumberOfGuests() {
        return numberOfGuests;
    }

    public boolean isOccupied() {
        return occupied;
    }

    public boolean isNotOccupied() {
        return !occupied;
    }
}

File: src/main/java/kitchenpos/eatinorders/tobe/domain/OrderTable.kt
package kitchenpos.eatinorders.tobe.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Table
import kitchenpos.eatinorders.tobe.domain.OrderTableGuestNumber.Companion.ZERO
import java.util.*

@Table(name = "order_table")
@Entity
class OrderTable private constructor(name: OrderTableName) {
    @Column(name = "id", columnDefinition = "binary(16)")
    @Id
    val id: UUID = UUID.randomUUID()

    @Column(name = "name", nullable = false)
    val name: OrderTableName = name

    @Column(name = "number_of_guests", nullable = false)
    var numberOfGuests: OrderTableGuestNumber = ZERO
        private set

    @Column(name = "occupied", nullable = false)
    var occupied: Boolean = false
        private set

    companion object {
        fun of(name: OrderTableName): OrderTable {
            return OrderTable(name)
        }
    }

    fun sit() {
        occupied = true
    }

    fun clear(orderTableClearValidator: OrderTableClearValidator) {
        orderTableClearValidator.validate(this)

        numberOfGuests = ZERO
        occupied = false
    }

    fun changeNumberOfGuest(numberOfGuests: OrderTableGuestNumber) {
        if (!occupied) {
            throw IllegalArgumentException("미점유 중인 테이블의 고객 좌석 수를 변경할 수 없습니다")
        }

        this.numberOfGuests = numberOfGuests
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as OrderTable

        return id == other.id
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

}
